### PR TITLE
Fix head-of-line blocking in SLC for subquery shapes via ETS link-values cache and inverted index

### PR DESCRIPTION
**Pull request by KyleAMathews**
Shapes with subqueries (\`IN (SELECT ...)\`) were causing 6–14s SLC stalls under load. This fixes three independent bottlenecks: O(N×D) serialised GenServer calls in the SLC filter step, O(N) other_shapes scan for dep-shape record changes, and O(D) ETS probe per shape in the hot-path membership check.

## Root Cause

Three separate hot paths were serialising work:

**Bottleneck 1 — SLC filter step (O(N×D×C) GenServer calls)**

\`get_link_values\` made a \`GenServer.call\` per invocation. In \`other_shapes_affected\`, this is called for every outer shape × every dep × every changed record. With N=300, D=3, C=20 that's ~18,000 sequential GenServer calls from a single process per transaction.

**Bottleneck 2 — Dep-shape record changes (O(N) other_shapes scan)**

When a dep-shape's own table changes (e.g. a row is inserted into the subquery's source table), \`Filter.affected_shapes\` had to scan all N outer shapes to find which ones were affected, evaluating each WHERE clause against the Materializer's current link values.

**Bottleneck 3 — Hot-path membership check (O(D) ETS probes per shape)**

\`registered_in_inverted_index?\` previously walked each shape's \`shape_dependencies_handles\` list and did an ETS lookup per dep handle to check membership — O(D) probes, called for every shape in \`other_shapes_affected\` on every change.

## Approach

**Fix 1: Per-stack ETS link-values cache**

Cache \`value_counts\` in a named ETS table (\`read_concurrency: true\`) owned by \`ConsumerRegistry\`. \`get_link_values\` does an ETS lookup (~5µs) instead of a GenServer call (~100µs + queue wait). Written after each committed transaction and after initial materialization.

**Fix 2: Inverted index in Filter**

Two new private ETS tables in \`Filter\` — \`sublink_field_table\` and \`sublink_dep_table\` — form an inverted index for dep shapes that land in non-optimisable WHERE clauses. When a dep-table record changes, the index directly returns the outer shapes whose link field matches, avoiding the O(N) scan entirely.

**Fix 3: O(1) membership check via in-struct MapSet**

Added \`sublink_shapes_set\` (a \`MapSet\` on the \`Filter\` struct) tracking which shape IDs are registered in the inverted index. \`registered_in_inverted_index?\` is now \`MapSet.member?\` — O(1) and zero ETS traffic — moved to run *before* the \`get_shape\` ETS lookup in \`other_shapes_affected\` so indexed shapes never touch ETS at all.

A guard (\`map_size(sublink_fields) > 0\`) ensures composite-PK subqueries like \`(user_id, team_id) IN (SELECT ...)\` — which produce no indexable fields — are never added to \`sublink_shapes_set\` and continue to be evaluated by \`other_shapes_affected\` normally. Without this guard they would be silently dropped.

Two new telemetry spans expose the remaining work: \`filter.sublink_candidates_count\` and \`filter.sublink_reeval.duration_µs\`.

**Key invariants:**
- ETS is written *after* each committed transaction — readers always see a consistent snapshot
- The inverted index is only built for dep shapes in top-level \`other_shapes\` (non-optimisable WHERE). Dep shapes that go through an equality index are already efficiently handled and are not registered here (preventing double-evaluation)
- \`sublink_affected_shapes\` re-evaluates the full WHERE clause for candidates to correctly handle compound conditions with non-sublink predicates
- Stale ETS entries are cleaned up in \`Consumer.terminate/2\` and the \`dependency_materializer_down\` handler
- If the link-values ETS table is missing (e.g. ConsumerRegistry restart window), \`sublink_affected_shapes\` returns an empty set rather than cascading to the expensive "return all shapes" fallback
- Composite-PK dep shapes (\`RowExpr\` left-hand sides) produce no indexable sublink fields and are never added to \`sublink_shapes_set\`; they correctly fall through to \`other_shapes_affected\`

**Non-goals:** This doesn't fix O(N) consumer fan-out in \`broadcast\` — only the O(D) Materializer serialisation within each consumer. Fixing fan-out would require a different pubsub model.

**Trade-offs:** ETS is write-once-per-txn, read-many. Stale reads between write and next commit are not a concern since \`get_link_values\` is only called during the SLC filter step, which runs *after* the commit that wrote the values.

## Verification

```sh
cd packages/sync-service

# Materializer ETS cache (43 tests)
mix test test/electric/shapes/consumer/materializer_test.exs

# ShapeLogCollector (24 tests)
mix test test/electric/shape_log_collector_test.exs

# Filter inverted index (covered by existing filter tests)
mix test test/electric/shapes/filter_test.exs

# Composite PK subquery regression
mix test test/electric/plug/router_test.exs:2671 --include slow
```

The regression test \`get_link_values reads from ETS cache and does not require the GenServer to be alive\` stops the Materializer GenServer after populating ETS, then asserts \`get_link_values\` returns the correct values — fails without the fix, passes with it.

## Files Changed

- **\`consumer/materializer.ex\`** — adds \`init_link_values_table/1\`, \`write_link_values/1\`, \`delete_link_values/2\`; updates \`get_link_values/1\` to check ETS first with GenServer fallback; adds logging when cache degrades
- **\`consumer_registry.ex\`** — creates the shared ETS table in \`new/2\` so it lives alongside the registry
- **\`consumer.ex\`** — calls \`delete_link_values\` in \`terminate/2\` and \`dependency_materializer_down\` handler to prevent stale cache entries
- **\`filter.ex\`** — adds inverted index ETS tables and \`sublink_shapes_set\` MapSet; \`register_sublink_shape/unregister_sublink_shape\` maintain both; \`sublink_affected_shapes\` does the O(K) lookup; \`registered_in_inverted_index?\` is now O(1); composite-PK guard prevents RowExpr shapes from being falsely indexed
- **\`filter/where_condition.ex\`** — moves \`registered_in_inverted_index?\` check before \`get_shape\` ETS lookup so indexed shapes are skipped without touching ETS
- **\`consumer/materializer_test.exs\`** — regression test for ETS cache
- **\`test/electric/shapes/filter_test.exs\`** — \`snapshot_filter_ets\` extended to cover both new ETS tables

---
Fixes https://github.com/electric-sql/electric/issues/3928

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by alco** [comment]
Not bad. Needs additional verification:

- we need a benchmark to validate the problem exists as described by the agent and that the changes resolve it
- switching from a synchronous GenServer.call to async ETS lookups need to be scrutinized for correctness in the specific code paths where `get_link_values` is called
- unleashing [Postgres oracle property tests](https://github.com/electric-sql/electric/pull/3935) on the fix could be a good way to gain confidence in the fix. But note that the oracle tests fail on `main` currently when a particular edge case is hit. We're tracking [one such edge case](https://github.com/electric-sql/electric/issues/3892), there may be others

File: packages/sync-service/lib/electric/replication/shape_log_collector.ex [hunk 1]
```diff
@@ -235,6 +235,8 @@ defmodule Electric.Replication.ShapeLogCollector do
         registry_state: registry_state
       })
 
+    :ets.new(:SLC_lookup_counter, [:named_table, :public, :set, write_concurrency: true])
```
Comment (icehaunter) [inline packages/sync-service/lib/electric/replication/shape_log_collector.ex]: This will not work in our cloud env, where this named table will have been taken by SLC of another tenant already.

**Review by icehaunter — DISMISSED** [review]

File: packages/sync-service/lib/electric/replication/shape_log_collector.ex (hunk 1 above)
Comment (alco) [reply packages/sync-service/lib/electric/replication/shape_log_collector.ex]: It's a temporary counter I added for benchmarking purposes. Sorry for the confusion.

**Comment by alco** [comment]
One thing Claude has highlighted:
>  Materializer crash: If a materializer crashes, its ETS entry persists (the table is owned by the SLC process via ConsumerRegistry). Stale values in ETS after a materializer crash are acceptable because:
>   - The materializer has restart: :temporary — it won't restart
>   - A crashed materializer triggers invalidation of dependent outer shapes
>   - The shape handle for the dead materializer won't be queried again

whilst Codex thinks it's a serious concern:
> • 1. High: stale ETS cache can suppress the existing safety fallback and cause false negatives in affected-shape routing.
>      get_link_values/1 now returns ETS data without checking materializer liveness (materializer.ex:76).
>      If a materializer dies, its ETS key is not cleared in terminate/down paths (materializer.ex:250, materializer.ex:270).
>      affected_shapes/3 relies on exceptions to fall back to “return all shapes” (where_condition.ex:167); stale ETS avoids that exception path and can make routing use old refs instead of conservative fan-out.
>      The new regression test explicitly enforces this dead-process ETS-read behavior (materializer_test.exs:407).
> 
>   Verdict: replacing GenServer.call with ETS lookups is valid for performance, but not fully correct as implemented unless you also guarantee cache invalidation/liveness coupling.
>   In steady state it is fine; under materializer death/restart races it can return stale refs and miss updates.
> 
>   I also ran:
> 
>   - mix test test/electric/shapes/consumer/materializer_test.exs test/electric/replication/shape_log_collector_test.exs (67 tests, 0 failures)
> 
>   Potential fix directions:
> 
>   1. delete ETS row on materializer shutdown/shape removal, or
>   2. include generation/version in ETS value and reject stale generations, or
>   3. on ETS hit, verify process alive for that handle before trusting cache.

**Comment by alco** [comment]
@claude 
Your suggestion for

> Stale ETS entries on shape deletion

to add a `terminate()` function won't work: the Materializer process doesn't trap exits, so its `terminate()` function won't be called.

Consumer process is the one acting on materializer's life cycle events (search for `:materializer_down` and `:dependency_materializer_down` in consumer.ex), so either it or the ShapeCleaner module must clean up the stale ETS entry.


---

Also, pay attention to [this comment](https://github.com/electric-sql/electric/pull/3937#issuecomment-3991919301).

**Comment by KyleAMathews** [comment]
## Latest commit: Fix stale ETS cleanup and tighten unregister guard

Addresses @alco's concern about stale ETS entries:

- Added `Materializer.delete_link_values/2` to remove ETS entries.
- Cleanup is now called in two places in `consumer.ex`:
  1. `terminate/2` — when a dep shape's consumer shuts down for any reason
  2. `handle_info({:dependency_materializer_down, handle})` — when a dep materializer crashes (before `handle_materializer_down` escalates to `stop_and_clean`)
- Also tightened `remove_shape` in `filter.ex` to only call `unregister_sublink_shape` when the shape was actually registered in the inverted index (via `registered_in_inverted_index?`).

---

**Re: CI failures** — all three reported failures (`FilterTest @>` performance, `ShapeStatusTest` high-concurrency, `ConsumerTest` txn fragments) are pre-existing issues unrelated to this PR:

- **FilterTest**: The reduction budget was calibrated for OTP 27. The repo's `.tool-versions` requests OTP 28.1.1 which has ~40% higher reduction counts for the same operations. Confirmed by checking out main branch's `filter.ex` — the test still fails identically. Passes on OTP 27.
- **ShapeStatusTest**: The slow test's 60s timeout is occasionally exceeded under CI load. Passes locally in 2.9s. The Walker appears in traces because `Shape.new/2` calls the parser, which uses Walker for protocol dispatch — this is not introduced by our changes.
- **ConsumerTest**: Flaky due to `Repatch.patch(:shared)` in concurrent tests.

File: packages/sync-service/lib/electric/shapes/consumer/materializer.ex [hunk 2]
```diff
@@ -57,11 +48,45 @@ defmodule Electric.Shapes.Consumer.Materializer do
...
+    case :ets.lookup(table, shape_handle) do
+      [{^shape_handle, values}] -> values
+      _ -> genserver_get_link_values(opts)
+    end
+  rescue
+    ArgumentError -> genserver_get_link_values(opts)
+  end
+
+  defp genserver_get_link_values(opts) do
+    GenServer.call(name(opts), :get_link_values, :infinity)
```
Comment (alco) [inline packages/sync-service/lib/electric/shapes/consumer/materializer.ex]: I don't understand why this call's timeout had to be changed to `:infinity`.

File: packages/sync-service/lib/electric/shapes/consumer/materializer.ex [hunk 3]
```diff
@@ -57,11 +48,45 @@ defmodule Electric.Shapes.Consumer.Materializer do
...
+  in a stack. Called by `ConsumerRegistry` during stack initialization. Idempotent —
+  safe to call when the table already exists.
+  """
+  @spec init_link_values_table(stack_id :: term()) :: :ets.table() | :undefined
+  def init_link_values_table(stack_id) do
+    :ets.new(link_values_table_name(stack_id), [
+      :named_table,
+      :public,
+      :set,
+      read_concurrency: true
```
Comment (alco) [inline packages/sync-service/lib/electric/shapes/consumer/materializer.ex]: Since this is a single shared table per stack, different materializer processes will be writing to it concurrently and the table itself becomes the new bottleneck. We need to add `write_concurrency: true` here and hope that'll alleviate lock contention.

**Comment by alco** [comment]
I'm rerunning my benchmarks on the latest version of the code.

**Comment by balegas** [comment]
This is wild. Well done on getting this merged so quickly.
